### ISSUE TITLE
Generalize SigAn Support

### ISSUE THREAD

**Issue by aromanielloNTIA** [NTIA/sigmf-ns-ntia#33]
## Issue

Signal analyzers have a variety of configurable parameters, depending on their specific hardware and their manufacturer’s design decisions. While many of these parameters are universal, such as a configured center frequency, others only apply to specific analyzers. For instance, some devices allow a configurable attenuation, while others use gain, and others use reference level. Currently, device-specific configuration parameters are left out of SigMF metadata since there is no general metadata field for these values.

## Solution

Capture all signal analyzer configuration parameters in metadata by further extending the `ntia-sensor` SigMF namespace extension. 

### Classes

To do this, ‘classes’ of signal analyzers are defined, as names for groups of signal analyzers which share the same configuration parameters. For example:

| Class | Devices in class |
|-------|---------------|
| `tekrsa_usb` | Tektronix RSA 300, 500, and 600 series devices |
| `usrp` |  USRP devices sharing the same config parameters |

### Extension of `ntia-sensor`'s `SensorAnnotation`

By adding these parameters to the SensorAnnotation, we are able to associate specific samples with specific configured sigan settings. Therefore, if a parameter is adjusted between two parts of the same data set, this is properly and easily tracked using the existing Annotation functionality. A new property is added to ntia-sensor’s SensorAnnotation:

| Name | Required | Type | Unit | Description |
|-------|-----------|-----|-------|-------------|
| `sigan_settings` | False | `SiganSettings` | N/A | Signal analyzer-specific device configuration parameters |

The new property is an instance of a new object, SiganSettings, which has only one required property:

| Name | Required | Type | Unit | Description |
|-------|-----------|-----|-------|-------------|
| `class` | True | String | N/A | A string matching a specified signal analyzer class |

A number of non-required properties are then defined and associated with each class as needed. For example, the tekrsa_usb class would define the following additional properties of SiganSettings:

| Name | Required | Type | Unit | Description |
|-------|-----------|-----|-------|-------------|
| `reference_level` | False | Double | dBm | Tektronix RSA USB reference level setting |
| `iq_bandwidth` | False | Double | Hz | Tektronix RSA USB IQ bandwidth setting |
| `attenuation` | False | Double | dB | Tektronix RSA 500A/600A series attenuation setting in manual attenuation mode |
| `preamp_enable` | False | Boolean | N/A | Tektronix RSA 500A/600A series preamp enable setting |
| `auto_attenuation_enable` | False | Boolean | N/A | Tektronix RSA 500A/600A series auto attenuation setting |

An example of this in use in a SigMF metadata file, for a measurement recorded with a Tektronix RSA 500A/600A series device:

```json
...
"annotations": [{
  "ntia-core:annotation_type": "SensorAnnotation",
  "core:sample_start": 0,
  "ntia-sensor:sigan_settings": {
    "class": "tekrsa_usb",
    "reference_level": -33.0,
    "iq_bandwidth": 10000000.0,
    "attenuation": 0.0,
    "preamp_enable": true,
    "auto_attenuation_enable": false
  }
}]

```

**Comment by aromanielloNTIA** [comment]
Update to the ntia-sensor spec available [here](https://github.com/aromanielloNTIA/sigmf-ns-ntia/tree/generalize-sigan-support).

The spec update is complete for the Tektronix RSA 3XX/5XX/6XX signal analyzer class, although more settings should likely be added to the USRP B2XX and Keysight N6841A classes. 

Due to removing the `SensorAnnotation` properties `attenuation_setting_sigan` and `gain_setting_sigan`, this update will be a breaking change to the spec. Those values are now recorded in the `SiganSettings` object.

**Comment by aromanielloNTIA** [comment]
Added `clock_rate` setting for USRP devices. This covers all settings configurable in the scos-usrp and scos-keysight plugins.

**Comment by aromanielloNTIA** [comment]
The format suggested here does not support recordings with multiple captures in which sigan settings vary between captures. I suggest keeping the `SiganSettings` object as defined here, but moving it into the `capture segement` objects in the `captures` array instead of being a part of the `SensorAnnotation`. This would implicitly associate the `SiganSettings` object with a set of samples corresponding to a single capture.

**Comment by aromanielloNTIA** [comment]
The implementation differs from what is described here, but this issue should be closed as resolved after #42 is merged, since it adds fields for notating sigan settings
